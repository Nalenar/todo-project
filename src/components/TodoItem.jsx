import { Icon } from "@iconify/react";

const TodoItem = (props) => (
  <div className="mx-3 mt-3 h-12 flex items-center shadow-md border border-gray-200 rounded last-of-type:mb-2">
    <div className="mx-3" onClick={() => props.handleComplete(props.todo.id)}>
      {props.todo.complete ? (
        <Icon icon="ion:checkmark-circle" width="27" color="#3b82f6" />
      ) : (
        <Icon icon="mdi:circle-outline" width="27" color="#3b82f6" />
      )}
    </div>
    <p
      className={`h-full w-full flex items-center overflow-x-auto text ${
        props.todo.complete ? "line-through text-gray-400 italic" : ""
      }`}
    >
      {props.todo.text}
    </p>
    <div
      className="h-[30px] w-[30px] flex justify-center items-center cursor-pointer mx-3"
      onClick={() => props.handleDelete(props.todo.id)}
    >
      <Icon icon="ion:trash" width={27} color="#ef4444" />
    </div>
  </div>
);

export default TodoItem;
