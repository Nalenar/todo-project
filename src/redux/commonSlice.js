/**
 * Redux слайс который содержит в себе всю логику связанную cо стандартными функциями
 * @module commonSlice
 * @function createSlice - создает слайс
 *
 * @param {string} name - Название
 * @param {array} initialState - Начальное состояние
 * @param {object} reducers - Редюсеры
 *
 * @param {object} state - состояние на данный момент
 * @param {object} action - данные действия
 *
 * @function toggleStyles - изменяет значение выбранного стиля
 * @function toggleWindows - изменяет значение выбранного окна
 * @function editTodo - передает id задачи в состояние
 */

import { createSlice } from "@reduxjs/toolkit";

export const commonSlice = createSlice({
  name: "common",
  initialState: {
    styles: {
      blur: "",
      theme: "light",
    },
    windows: {
      menu: false,
      settings: false,
      modal: false,
      editList: false,
      editTodo: false,
    },
    edit: {
      id: undefined,
    },
    search: undefined,
  },
  reducers: {
    toggleStyles: (state, action) => {
      state.styles[action.payload.style] = action.payload.string;
    },
    toggleWindows: (state, action) => {
      // if (action.payload.window === "menu") {
      //   state.windows = {
      //     menu: action.payload.toggle,
      //     settings: false,
      //     modal: false,
      //     editList: false,
      //     editTodo: false,
      //   };
      // } else if (action.payload.window === "settings") {
      //   state.windows = {
      //     menu: false,
      //     settings: action.payload.toggle,
      //     modal: false,
      //     editList: false,
      //     editTodo: false,
      //   };
      // } else {
      //   state.windows[action.payload.window] = action.payload.toggle;
      // }
      state.windows[action.payload.window] = action.payload.toggle;
    },
    editTodo: (state, action) => {
      state.edit.id = action.payload.id;
    },
    searchTodo: (state, action) => {
      state.search = action.payload.value;
    },
  },
});

/** @exports commonSlice.actions  - возможные действия со стандартными функциями */
export const { toggleStyles, toggleWindows, editTodo, searchTodo } =
  commonSlice.actions;

/**
 * @default
 * @exports commonSlice.reducer - экспортирует по-умолчанию редюсер всего слайса
 */
export default commonSlice.reducer;
