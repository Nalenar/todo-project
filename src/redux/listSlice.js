/**
 * Redux слайс который содержит в себе всю логику связанную cо списками задач
 * @module listSlice
 * @function createSlice - создает слайс
 *
 * @param {string} name - Название
 * @param {array} initialState - Начальное состояние
 * @param {object} reducers - Редюсеры
 *
 * @param {object} state - состояние на данный момент
 * @param {object} action - данные действия
 *
 * @function addCategory - создает список
 * @function deleteCategory - удаляет список
 * @function renameCategory - изменяет название списка на новое, указанное пользователем
 */

import { createSlice } from "@reduxjs/toolkit";

export const listSlice = createSlice({
  name: "lists",
  initialState: [
    {
      id: "HELP ME",
      title: "HELP ME",
    },
  ],
  reducers: {
    addCategory: (state, action) => {
      const list = {
        id: action.payload.title,
        title: action.payload.title,
      };
      state.push(list);
    },
    deleteCategory: (state, action) => {
      return state.filter((list) => list.id !== action.payload.id);
    },
    renameCategory: (state, action) => {
      const index = state.findIndex((list) => list.id === action.payload.id);
      state[index].title = action.payload.title;
    },
  },
});

/** @exports listSlice.actions  - возможные действия со списками */
export const { addCategory, deleteCategory, renameCategory } =
  listSlice.actions;

/**
 * @default
 * @exports listSlice.reducer - экспортирует по-умолчанию редюсер всего слайса
 */
export default listSlice.reducer;
