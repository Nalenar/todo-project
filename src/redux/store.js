/**
 * Главный файл который использует store
 * из библиотеки @reduxjs/toolkit
 * @module
 * 
 * Экспортирует объект хранящий в себе редюсеры слайсов
 * @exports 
 * 
 * @param {object} reducer - Содержит в себе название состояния и относящийся к нему редюсер
 * 
 * @param {object} todos - Содержит в себе все экшены, редюсеры и состояния для задач
 * @param {object} lists - Содержит в себе все экшены, редюсеры и состояния для списков 
 * @param {object} common - Содержит в себе все экшены, редюсеры и состояния для стандартных функций (например стилей, окон и т.д.)
 */

import { configureStore } from "@reduxjs/toolkit";
import todoReducer from "./todoSlice";
import commonReducer from "./commonSlice";
import listReducer from "./listSlice";

export default configureStore({
  reducer: {
    todos: todoReducer,
    lists: listReducer,
    common: commonReducer,
  },
});
