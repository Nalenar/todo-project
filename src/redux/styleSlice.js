import { createSlice } from "@reduxjs/toolkit";

export const styleSlice = createSlice({
  name: "blur",
  initialState: {
    blur: "",
  },
  reducers: {
    addBlur: (state) => {
      state.blur = "blur-sm";
    },
    deleteBlur: (state) => {
      state.blur = "";
    },
  },
});

export const { addBlur, deleteBlur } = styleSlice.actions;
export default styleSlice.reducer;
