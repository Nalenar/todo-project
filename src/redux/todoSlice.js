/**
 * Redux слайс который содержит в себе всю логику связанную с задачами
 * @module todoSlice
 * @function createSlice - Создает слайс
 *
 * @param {string} name - Название
 * @param {array} initialState - Начальное состояние
 * @param {object} reducers - Редюсеры
 *
 *
 * @param {object} state - состояние на данный момент
 * @param {object} action - данные действия
 *
 * @function addTodo - добавляет задание(объект) в состояние
 * @function toggleComplete - находит нужное задание меняет заначение completed
 * @function deleteTodo - удаляет нужное задание
 * @function toggleList - удаляет или добавляет список в состоянии
 */

import { createSlice } from "@reduxjs/toolkit";

export const todoSlice = createSlice({
  name: "todos",
  initialState: [
    {
      id: 1,
      title: "today",
      completed: true,
      list: ["today"],
    },
    {
      id: 2,
      title: "important",
      completed: true,
      list: ["today", "important"],
    },
  ],
  reducers: {
    addTodo: (state, action) => {
      const todo = {
        id: state.length + 1,
        title: action.payload.title,
        completed: false,
        list: [action.payload.list],
      };
      state.push(todo);
    },
    toggleComplete: (state, action) => {
      const index = state.findIndex((todo) => todo.id === action.payload.id);
      state[index].completed = action.payload.completed;
    },
    deleteTodo: (state, action) => {
      return state.filter((todo) => todo.id !== action.payload.id);
    },
    toggleList: (state, action) => {
      const index = state.findIndex((todo) => todo.id === action.payload.id);
      if (action.payload.toggle) {
        state[index].list.push(`${action.payload.list}`);
      } else {
        state[index].list = state[index].list.filter(
          (list) => list !== action.payload.list
        );
      }
    },
  },
});

/** @exports todoSlice.actions - экспортирует действия */

export const { addTodo, toggleComplete, deleteTodo, toggleList } =
  todoSlice.actions;

/**
 * @default
 *  @exports todoSlice.reducer - экспортирует по-умолчанию редюсер всего слайса
 */
export default todoSlice.reducer;
