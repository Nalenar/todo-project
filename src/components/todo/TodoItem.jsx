import { Icon } from "@iconify/react";
import { useDispatch } from "react-redux";
import { toggleComplete, deleteTodo } from "../../redux/todoSlice";

const TodoItem = (props) => {
  const dispatch = useDispatch();

  const handleCompleteClick = () => {
    dispatch(
      toggleComplete({ id: props.todo.id, completed: !props.todo.completed })
    );
  };

  const handleDeleteClick = () => {
    dispatch(deleteTodo({ id: props.todo.id }));
  };

  return (
    <div className="mx-3 mt-3 h-12 flex items-center shadow-md border border-gray-200 rounded last-of-type:mb-2">
      <div className="mx-3 cursor-pointer" onClick={handleCompleteClick}>
        {props.todo.completed ? (
          <Icon icon="ion:checkmark-circle" width="27" color="#3b82f6" />
        ) : (
          <Icon icon="mdi:circle-outline" width="27" color="#3b82f6" />
        )}
      </div>
      <p
        className={`h-full w-full flex items-center overflow-x-auto text ${
          props.todo.completed ? "line-through text-gray-400 italic" : ""
        }`}
      >
        {props.todo.title}
      </p>
      <div
        onClick={handleDeleteClick}
        className="h-[30px] w-[30px] flex justify-center items-center cursor-pointer mx-3"
      >
        <Icon icon="ion:trash" width={27} color="#ef4444" />
      </div>
    </div>
  );
};

export default TodoItem;
