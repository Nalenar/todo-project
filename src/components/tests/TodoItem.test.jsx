import { it, expect, describe, vi } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import * as reduxHooks from "react-redux";

import * as todoActions from "../../redux/todoSlice";
import * as commonActions from "../../redux/commonSlice";

import TodoItem from "../todo/TodoItem";

vi.mock("react-redux");

const todo = {
  id: 1,
  title: "today",
  completed: true,
  list: ["today"],
};

const mockedDispatch = vi.spyOn(reduxHooks, "useDispatch");

describe("TodoItem", () => {
  it("should create TodoItem", () => {
    mockedDispatch.mockReturnValue(vi.fn());

    const component = render(<TodoItem todo={todo} />);

    expect(component).toMatchSnapshot();
  });

  it("should dispatch actions", () => {
    const dispatch = vi.fn();
    mockedDispatch.mockReturnValue(dispatch);

    const mockedToggleComplete = vi.spyOn(todoActions, "toggleComplete");
    const mockedToggleImportant = vi.spyOn(todoActions, "toggleList");
    const mockedToggleWindows = vi.spyOn(commonActions, "toggleWindows");
    const mockedEditTodo = vi.spyOn(commonActions, "editTodo");

    render(<TodoItem todo={todo} />);

    fireEvent.click(screen.getByRole("button-complete"));

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(mockedToggleComplete).toHaveBeenCalledWith({
      id: todo.id,
      completed: !todo.completed,
    });

    fireEvent.click(screen.getByRole("button-important"));

    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(mockedToggleImportant).toHaveBeenCalledWith({
      id: todo.id,
      list: "important",
      toggle: !todo.list.includes("important"),
    });

    fireEvent.click(screen.getByRole("todo-item"));

    expect(dispatch).toHaveBeenCalledTimes(4);
    expect(mockedToggleWindows).toHaveBeenCalledWith({
      window: "editTodo",
      toggle: true,
    });
    expect(mockedEditTodo).toHaveBeenCalledWith({ id: todo.id });
  });
});
