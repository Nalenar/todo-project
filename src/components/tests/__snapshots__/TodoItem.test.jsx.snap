// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoItem > should create TodoItem 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mx-3 mt-3 flex h-12 cursor-pointer items-center rounded border border-gray-200 bg-white shadow-md last-of-type:mb-2"
      >
        <div
          class="mx-3 cursor-pointer"
          role="button-complete"
        >
          <span />
        </div>
        <p
          class="flex h-full w-full items-center overflow-x-auto italic text-gray-400 line-through"
          role="todo-item"
        >
          today
        </p>
        <button
          class="mx-3 flex h-[30px] w-[30px] cursor-pointer items-center justify-center"
          role="button-important"
        >
          <span />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mx-3 mt-3 flex h-12 cursor-pointer items-center rounded border border-gray-200 bg-white shadow-md last-of-type:mb-2"
    >
      <div
        class="mx-3 cursor-pointer"
        role="button-complete"
      >
        <span />
      </div>
      <p
        class="flex h-full w-full items-center overflow-x-auto italic text-gray-400 line-through"
        role="todo-item"
      >
        today
      </p>
      <button
        class="mx-3 flex h-[30px] w-[30px] cursor-pointer items-center justify-center"
        role="button-important"
      >
        <span />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
